pkg/services/kubeconfig_loader.go

```diff
@@ -0,0 +1,239 @@
+	}
+	maxRetries, err := strconv.Atoi(getEnv("MAX_RETRIES", "60"))
+	if err != nil {
+		logger.Errorf("Error parsing MAX_RETRIES environment variable: %s", err)
+		return nil
+	}
+	return &KubeconfigLoader{
+		path:              kubeConfigPath,
+		logger:            logger,
+		watcher:           watcher,
+		remoteKubeClients: remoteKubeClients,
+		lastConfigHash:    "",
+		clientFactory:     nil,
+		maxRetries:        maxRetries,
+	}
+}
+
+// Cleanup closes the watcher and logs an error if it fails.
+func (k *KubeconfigLoader) Cleanup() {
+	defer k.logger.Info("Kubeconfig watcher closed")
+	if err := k.watcher.Close(); err != nil { // fsnotify watcher Linux-backend Close() always returns nil, but we check for errors to be safe
+		k.logger.Errorf("Error closing watcher: %s", err)
+	}
+}
+
+// hasKubeconfigChanged checks if the kubeconfig file has changed by comparing its hash.
+func (k *KubeconfigLoader) hasKubeconfigChanged() (bool, string, error) {
+	content, err := os.ReadFile(k.path)
+	if err != nil {
+		return false, "", fmt.Errorf("error reading kubeconfig file: %w", err)
+	}
+
+	hash := fmt.Sprintf("%x", sha256.Sum256(content))
+
+	return hash != k.lastConfigHash, hash, nil
+}
+
+// LoadKubeconfig loads the kubeconfig from the filesystem and initializes remote clusters.
+func (k *KubeconfigLoader) LoadKubeconfig() {
+	changed, currentHash, err := k.hasKubeconfigChanged()
+	if err != nil {
+		k.logger.Errorf("Failed to check kubeconfig changes: %s", err)
+		return
+	}
+	if !changed {
+		k.logger.Debugf("No changes detected in kubeconfig %s, skipping reload", k.path)
+		return
+	}
+	k.logger.Debugf("Detected kubeconfig change, reloading: %s", k.path)
+	for client := range k.remoteKubeClients { // if we reassign the map, the synchronizers lose the reference to it
+		delete(k.remoteKubeClients, client)
+		k.logger.Debugf("Removed remote client for cluster: %s", client)
+	}
+
+	// Update stored hash
+	k.lastConfigHash = currentHash
+
+	kubeConfig, err := clientcmd.LoadFromFile(k.path)
+	if err != nil { // kubeConfig is nil only if the file is empty or invalid
+		k.logger.Errorf("Error loading kube config from path %s: %s", k.path, err)
+		return
+	}
+	if len(kubeConfig.Contexts) == 0 {
+		k.logger.Warnf("No contexts found in kubeconfig file: %s", k.path)
+		return
+	}
+
+	var remoteClustersName []string
+	for contextName := range kubeConfig.Contexts {
+		remoteClustersName = append(remoteClustersName, contextName)
+	}
+	k.logger.Debugf("Remote clusters found in kubeconfig: %v", remoteClustersName)
+
+	if len(remoteClustersName) > 0 {
+		var initializedClustersName []string
+		for _, cluster := range remoteClustersName {
+			remoteClusterConfig, err := BuildConfigWithContextFromFlags(cluster, k.path)
+			if err != nil {
+				k.logger.Errorf("Error building kubeconfig for cluster '%s': %s", cluster, err)
+				continue
+			}
+
+			remoteClusterConfig.Timeout = 1 * time.Second // Set a timeout for the HTTP client, maybe this should be configurable
```

Comment:
Considering we already made max_retries configurable, we could maybe add a default value and make it optional, but configurable

Reply:
Fixed! Added a configurable flag parameter.